### ISSUE TITLE
Metadata in pallet-assets cannot be accessed by other pallet

### ISSUE THREAD

**Issue by MrPai** [paritytech/substrate#9877]
Hi, I have a question about [**pallet-assets**](https://github.com/paritytech/substrate/tree/master/frame/assets).
we are developing a DeFi project that need to use asset's decimal when calculating, however when we import **pallet-assets** in our project, the storage [Metadata](https://github.com/paritytech/substrate/blob/polkadot-v0.9.9/frame/assets/src/lib.rs#L267) can not be accessed by other pallets.
```
#[pallet::storage]
/// Metadata of an asset.
pub(super) type Metadata<T: Config<I>, I: 'static = ()> = StorageMap<
        _,
        Blake2_128Concat,
        T::AssetId,
        AssetMetadata<DepositBalanceOf<T, I>, BoundedVec<u8, T::StringLimit>>,
        ValueQuery,
        GetDefault,
        ConstU32<300_000>,
>;
```
I wonder can we make it public or abstract a trait, so the other pallet can access its decimal?

**Comment by bkchr** [comment]
CC @shawntabrizi @apopiak

**Comment by MrPai** [comment]
Yes, read access is enough, it seems like `decimal` can not be read even add `#[pallet::getter(fn metadata)]`, because of the type [AssetMetadata](https://github.com/paritytech/substrate/blob/polkadot-v0.9.9/frame/assets/src/types.rs#L90-L104).

```
#[derive(Clone, Encode, Decode, Eq, PartialEq, Default, RuntimeDebug, MaxEncodedLen)]
pub struct AssetMetadata<DepositBalance, BoundedString> {
	/// The balance deposited for this metadata.
	///
	/// This pays for the data stored in this struct.
	pub(super) deposit: DepositBalance,
	/// The user friendly name of this asset. Limited in length by `StringLimit`.
	pub(super) name: BoundedString,
	/// The ticker symbol for this asset. Limited in length by `StringLimit`.
	pub(super) symbol: BoundedString,
	/// The number of decimals this asset uses to represent one unit.
	pub(super) decimals: u8,
	/// Whether the asset metadata may be changed by a non Force origin.
	pub(super) is_frozen: bool,
}
```

**Comment by MrPai** [comment]
Now we define the decimal with hardcode in runtime to solve the problem, but this is error-prone.
our project is going to be a parachain soon, please help us with this if possible, thanks.

**Comment by shawntabrizi** [comment]
@MrPai in what way are you using the decimal information in the runtime?

In general this sounds really bad because the runtime should never be doing floating point logic. The decimal, and all of the metadata should be used exclusively for front-end experiences outside of the blockchain.

**Comment by MrPai** [comment]
Hi, @shawntabrizi

We have not done any floating-point logic on-chain. we used `FixedU128` instead. the reason why we need decimal is that in our **money market** pallet, we need to calculate the total amount by price * amount. 

As you know, USDT may have 6 decimal while KSM has 12, and DOT has 10. 
Well when on-chain, the `amount` looks like this: 
 - **one** USDT equals 1_000_000, 
 - **one** KSM equals 1_000_000_000_000,
 - **one** DOT equals 10_000_000_000,


when oracle feed price to on-chain, the `price` looks like this:
 - USDT/USD : 1.00064
 - DOT/USD   : 27.357
 - KSM/USD  : 331.41
and all the price wrap in `FixedU128`, so on-chain, it looks like:
 - USDT/USD : FixedU128(1,000,640,000,000,000,000)
 - DOT/USD   : FixedU128(27,357,815,270,000,009,216)
 - KSM/USD  : FixedU128(331,419,885,270,000,009,216)

**IMPORTANT**
From the above data, we cannot directly use **price * amount** due to the decimal difference, our formula looks like `(price/decimal) * amount`.

**This is the reason why we need the decimal on-chain.**

Can you help solve this requirement, we just change from `orml-currency` to `pallet-assets`, please let us know if you have any advice, thank you.

**Comment by shawntabrizi** [comment]
I have opened a simple PR to help you here, but in general, hard-coding things might not be such a bad idea. For example, how in general can you determine which asset corresponds to KSM, DOT, and USDT?

At that level, you must program that Asset ID X maps to some specific asset you care about, and in that case, it probably makes sense to hardcode all of the metadata about that asset in the interface. But anyway, hopefully that PR will solve things the way you invision.

**Comment by MrPai** [comment]
Yes, the PR helps a lot, we are thinking about dynamically specifying and associate assets in our own pallet. and it will be easy for us to maintain decimal only in one place, `pallet_assets` here. No need to care about decimal at two places(runtime and pallet-assets).
A big thanks for your help :)
